r3bbase/R3BTprevTnext.cxx

```diff
@@ -0,0 +1,170 @@
+        return kFATAL;
+    }
+
+    fR3BEventHeader = dynamic_cast<R3BEventHeader*>(rootManager->GetObject("EventHeader."));
+    if (fR3BEventHeader == nullptr)
+    {
+        LOG(error) << "R3BTprevTnext::Init() EventHeader. not found";
+        return kFATAL;
+    }
+
+    fSamplerMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMapped"));
+    if (fSamplerMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fSamplerMSMapped = dynamic_cast<TClonesArray*>(rootManager->GetObject("SamplerMSMapped"));
+    if (fSamplerMSMapped == nullptr)
+    {
+        return kFATAL;
+    }
+
+    fMSOffsetPar = dynamic_cast<R3BMSOffsetPar*>(rtdb->getContainer("MSOffsetPar"));
+    if (fMSOffsetPar == nullptr)
+    {
+        LOG(error) << "Could not find MSOffsetPar container!";
+        return kFATAL;
+    }
+
+    return kSUCCESS;
+}
+
+void R3BTprevTnext::Exec(Option_t* /*opt*/)
+{
+
+    fTprev = -1.0e3; // Initializaton of Tprev and Tnext
+    fTnext = -1.0e3; // If in the end you have this value is because this is an offspill or SAMP and/or SAMPMS are not
+                     // properly recorded
+    const Int_t sampHits = fSamplerMapped->GetEntriesFast();
+    const Int_t sampmsHits = fSamplerMSMapped->GetEntriesFast();
+    if ((sampHits < 1) || (sampmsHits != 1)) // If statement to check if the event is properly recorded
+    {
+        fR3BEventHeader->SetTprev(fTprev);
+        fR3BEventHeader->SetTnext(fTnext);
+        return;
+    }
+    const Double_t MSOffset = fMSOffsetPar->GetMSOffset();
+    Double_t SAMPTime = 0;
+    Double_t SAMPMSTime = 0;
+    R3BSamplerMappedData* SAMPMapped = nullptr;
+    R3BSamplerMappedData* SAMPMSMapped = nullptr;
+
+    SAMPMSMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMSMapped->At(0));
+    SAMPMSTime = SAMPMSMapped->GetTime();
+
+    if (fR3BEventHeader->GetTpat() < fTpat && fR3BEventHeader->GetTrigger() == 1 && fR3BEventHeader->GetTpat() > 0)
+    {
+        if (sampHits == 1)
+        {                                      //
+            fTprev = 1e8;                      // If statement to check if your valid event
+            fTnext = 1e8;                      // have only 1 hit
+            fR3BEventHeader->SetTprev(fTprev); // If yes, Tprev and Tnext will have value 1e8
+            fR3BEventHeader->SetTnext(fTnext); //
+            return;
+        }
+        fTprev = 1e9; // Initialization with a different value after the first tests
+        fTnext = 1e9; // If you have this value in your tprev/tnext is because the previous or next hit is inside the
+                      // delta_clk range but is not a MS
+        for (Int_t i = 0; i < sampHits; ++i)
+        {
+            SAMPMapped = dynamic_cast<R3BSamplerMappedData*>(fSamplerMapped->At(i));
+            SAMPTime = SAMPMapped->GetTime();
+            auto tpn = SAMPMSTime + MSOffset - SAMPTime;
+            if (tpn > delta_clk)
+            {
+                fTprev = tpn * 10.0; // in ns
```

Comment:
What is the reason behind this magic number 10.0?

Reply:
Yes. It's correct

Reply:
The trloii sampler counts clock cycles of a 100MHz clock. 
That means that one clock cycle is 10ns. 
(We have discussed using a uniform unit system before, but I don't think this is the correct place to start with that.)

Reply:
Again @klenze explained it

Reply:
In this case, maybe define a named constant on the top:
```cpp
constexpr auto CLOCK_PERIOD = 10.; // ns
```
And then use this variable in the code:
```
fTprev = tpn * CLOCK_PERIOD;
```